api/transport/header.go

```diff
@@ -65,16 +67,23 @@ func (h Headers) With(k, v string) Headers {
 	if h.items == nil {
 		h.items = make(map[string]string)
 	}
+	if h.keyMapping == nil {
+		h.keyMapping = make(map[string]string)
+	}
 
-	h.items[CanonicalizeHeaderKey(k)] = v
+	canonicalizedKey := CanonicalizeHeaderKey(k)
+	h.items[canonicalizedKey] = v
+	h.keyMapping[canonicalizedKey] = k
```

Comment:
Using a single map to struct will perform better for single header reads and writes, though not as well for bulk reads.